framework/src/modules/keys/keys_module.ts

```diff
@@ -38,4 +52,172 @@ export class KeysModule extends BaseModule {
 			numberOfSignatures: 0,
 		},
 	};
+
+	// eslint-disable-next-line class-methods-use-this
+	public async beforeTransactionApply({
+		stateStore,
+		transaction: {
+			asset,
+			assetType,
+			getSigningBytes,
+			id,
+			moduleType,
+			senderID,
+			senderPublicKey,
+			signatures,
+		},
+	}: TransactionApplyInput): Promise<void> {
+		const sender = await stateStore.account.get<AccountKeyAsset>(senderID);
+		const { networkIdentifier } = stateStore.chain;
+		const transactionBytes = getSigningBytes();
+
+		const transactionWithNetworkIdentifierBytes = Buffer.concat([
+			networkIdentifier,
+			transactionBytes,
+		]);
+
+		// This is for registration of multisignature that requires all signatures
+		if (moduleType === this.type && assetType === RegisterAssetType) {
+			const { mandatoryKeys, optionalKeys } = codec.decode<DecodedAsset>(KeysSchema, asset);
+
+			// For multisig registration we need all signatures to be present
+			const numberOfExpectedKeys = mandatoryKeys.length + optionalKeys.length + 1;
+			if (numberOfExpectedKeys !== signatures.length) {
+				throw new Error(
+					`There are missing signatures. Expected: ${numberOfExpectedKeys} signatures but got: ${signatures.length}`,
+				);
+			}
+
+			// Check if empty signatures are present
+			if (!signatures.every(signature => signature.length > 0)) {
+				throw new Error('A signature is required for each registered key.');
+			}
+
+			// Verify first signature is from senderPublicKey
+			validateSignature(senderPublicKey, signatures[0], transactionWithNetworkIdentifierBytes, id);
+
+			// Verify each mandatory key signed in order
+			validateKeysSignatures(
+				mandatoryKeys,
+				signatures.slice(1, mandatoryKeys.length + 1),
+				transactionWithNetworkIdentifierBytes,
+				id,
+			);
+
+			// Verify each optional key signed in order
+			validateKeysSignatures(
+				optionalKeys,
+				signatures.slice(mandatoryKeys.length + 1),
+				transactionWithNetworkIdentifierBytes,
+				id,
+			);
+			return;
+		}
+
+		if (!isMultisignatureAccount(sender)) {
+			validateSignature(senderPublicKey, signatures[0], transactionWithNetworkIdentifierBytes, id);
```

Comment:
was there any check for length of signatures in this case? i think `signatures` length should be exactly 1?

Reply:
Code was moved as it was.